### PR TITLE
HPCC-23075 XML Spray loses header length if there is not footer present.

### PR DESCRIPTION
**Pull request by AttilaVamos**
Fix header size loosing bug.

Add regression test for
- Check xml header and footer is right in size when dataset write out as XML file (HPCC-23079)
- Check the despray doesn't lose any record if they are smaller than headers (HPCC-23046)
- Check the spray creates same file as the original was with proper header/footer size

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Tested manually

### CODE REVIEW

**Comment by AttilaVamos** [comment]
@jakesmith please review.

File: dali/ft/filecopy.cpp [hunk 1]
```diff
@@ -1757,7 +1757,21 @@ void FileSprayer::derivePartitionExtra()
 void FileSprayer::displayPartition()
 {
     ForEachItemIn(idx, partition)
+    {
         partition.item(idx).display();
+
+#ifdef _DEBUG
+        if ((partition.item(idx).whichInput >= 0) && (partition.item(idx).whichInput < sources.ordinality()) )
+            LOG(MCdebugInfoDetail, unknownJob,
+                     "   Header size: %d, XML header size: %" I64F "d, XML footer size: %" I64F "d",
```
Comment (jakesmith) [inline dali/ft/filecopy.cpp]: trivial: all are unsigned so should be %u.

Incidentally why is headerSize an unsigned but xmlHeaderLength a offset_t - should be consistent I'd have thought?

File: dali/ft/filecopy.cpp [hunk 2]
```diff
@@ -3000,26 +3014,22 @@ void FileSprayer::spray()
 bool FileSprayer::isSameSizeHeaderFooter()
```
Comment (jakesmith) [inline dali/ft/filecopy.cpp]: what is the role of isSameSizeHeaderFooter() ?
I see some code that sets FPheaderLength and FPfooterLength as meta properties of each part:
```
                if (!sameSizeHeaderFooter)
                {
                    FilePartInfo & curHeaderSource = sources.item(whichHeaderInput);
                    curProps.setPropInt64(FPheaderLength, curHeaderSource.xmlHeaderLength);

                    FilePartInfo & curFooterSource = sources.item(cur.whichInput);
                    curProps.setPropInt64(FPfooterLength, curFooterSource.xmlFooterLength);

                    if ( idx+1 != partition.ordinality() )
                        whichHeaderInput = partition.item(idx+1).whichInput;
                }
```

so if not same as 1st part, then sets into part,
but I don't think I see where these per part properties are used?

File: dali/ft/filecopy.cpp [hunk 3]
```diff
@@ -3000,26 +3014,22 @@ void FileSprayer::spray()
...
     if (sources.ordinality() == 0)
         return retVal;
 
+    unsigned whichHeaderInput = 0;
+
+    headerSize = sources.item(whichHeaderInput).xmlHeaderLength;
+    footerSize = sources.item(whichHeaderInput).xmlFooterLength;
+
+
+
```
Comment (jakesmith) [inline dali/ft/filecopy.cpp]: trivial: a few too many blank lines

File: dali/ft/filecopy.cpp [hunk 4]
```diff
@@ -3000,26 +3014,22 @@ void FileSprayer::spray()
...
         PartitionPoint & cur = partition.item(idx);
         if (cur.inputLength && (idx+1 == partition.ordinality() || partition.item(idx+1).whichOutput != cur.whichOutput))
         {
-            if (isEmpty)
-            {
-                headerSize = sources.item(whichHeaderInput).xmlHeaderLength;
-                footerSize = sources.item(cur.whichInput).xmlFooterLength;
-                isEmpty = false;
-                continue;
-            }
```
Comment (jakesmith) [inline dali/ft/filecopy.cpp]: trivial: unnecessary blank line.

**Comment by jakesmith** [comment]
@AttilaVamos - a few trivial comments + 1 question.

**Comment by AttilaVamos** [comment]
@jakesmith updated.

File: dali/ft/filecopy.cpp (hunk 2 above)
Comment (jakesmith) [reply dali/ft/filecopy.cpp]: @AttilaVamos - was there an answer to this question?

File: dali/ft/filecopy.cpp (hunk 1 above)
Comment (jakesmith) [reply dali/ft/filecopy.cpp]: >trivial: all are unsigned so should be %u.

@AttilaVamos - looks like this one hasn't been addressed.

**Comment by jakesmith** [comment]
@AttilaVamos - there 1 trivial issue and a question that wasn't addressed/answered from before.

File: dali/ft/filecopy.cpp (hunk 2 above)
Comment (AttilaVamos) [reply dali/ft/filecopy.cpp]: It seems my previous answer went missing. 
This implementation was part of the https://track.hpccsystems.com/browse/HPCC-13331 more than 4 years ago. I think only the despray uses this information.

File: dali/ft/filecopy.cpp (hunk 1 above)
Comment (AttilaVamos) [reply dali/ft/filecopy.cpp]: It has been changed in my last commit/push.

**Comment by AttilaVamos** [comment]
@jakesmith The trivial issue already addressed in my last commit/push and the question answered.

**Comment by jakesmith** [comment]
>trivial: all are unsigned so should be %u.

I still see the %d in the PR/branch, see: https://github.com/hpcc-systems/HPCC-Platform/blob/bf9af6ee4ca35b98437ad8d6b8d12e637586dc44/dali/ft/filecopy.cpp

@AttilaVamos

File: dali/ft/filecopy.cpp (hunk 2 above)
Comment (jakesmith) [reply dali/ft/filecopy.cpp]: Ok thanks.

**Comment by AttilaVamos** [comment]
@jakesmith updated, please recheck.

**Comment by jakesmith** [comment]
Looks good.
@richardkchapman - ready to merge.

**Pull request merged**